File: zantt-api-backend/src/main/java/com/zantt/controllers/ProjectApiController.java
package com.zantt.controllers;

import com.zantt.mappers.ProjectMapper;
import com.zantt.models.AddProjectReqModel;
import com.zantt.models.ApiResponse;
import com.zantt.models.ProjectViewModel;
import com.zantt.services.ProjectService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/project")
public class ProjectApiController {
    @Autowired
    ProjectService projectService;

    @PostMapping("")
    public ResponseEntity<ApiResponse<ProjectViewModel>> addProject(@RequestBody AddProjectReqModel req) {
        var project = projectService.addProject(req.getName());
        var result = ProjectMapper.Instance.toViewModel(project);
        var response = new ApiResponse<>(result);
        return ResponseEntity.ok(response);
    }

    @GetMapping("/projects")
    public ResponseEntity<ApiResponse<List<ProjectViewModel>>> getProjects() {
        var projects = projectService.getProjects();
        var result = ProjectMapper.Instance.toViewModels(projects);
        var response = new ApiResponse<>(result);
        return ResponseEntity.ok(response);
    }

    @GetMapping("")
    public ResponseEntity<ApiResponse<ProjectViewModel>> getProject(@RequestParam("projectId") String projectId) {
        var project = projectService.getProject(projectId);
        var result = ProjectMapper.Instance.toViewModel(project);
        var response = new ApiResponse<>(result);
        return ResponseEntity.ok(response);
    }

    @DeleteMapping("")
    public ResponseEntity<ApiResponse<Object>> deleteProject(@RequestParam("projectId") String projectId) {
        projectService.deleteProject(projectId);
        var response = new ApiResponse<>(null);
        return ResponseEntity.ok(response);
    }
}

File: zantt-api-backend-net/Zantt/Controllers/ProjectApiController.cs
using AutoMapper;
using Microsoft.AspNetCore.Mvc;
using Zantt.Entities;
using Zantt.Exceptions;
using Zantt.Filters;
using Zantt.Models.RequestModels;
using Zantt.Models.ViewModels;
using Zantt.Services;

namespace Zantt.Controllers;

[ServiceFilter(typeof(ApiExceptionFilterAttribute))]
[Route("project")]
[ApiController]
public class ProjectApiController : ControllerBase
{
    private readonly ILogger<ProjectApiController> logger;
    private readonly ProjectService projectService;
    private readonly IMapper mapper;

    public ProjectApiController(
        ILogger<ProjectApiController> logger,
        ProjectService projectService,
        IMapper mapper)
    {
        this.logger = logger;
        this.projectService = projectService;
        this.mapper = mapper;
    }

    [HttpGet]
    [Route("projects")]
    public ApiResponse<IEnumerable<ProjectViewModel>> GetProjects()
    {
        var projects = projectService.GetProjects();
        var result = mapper.Map<List<ProjectEntity>, IEnumerable<ProjectViewModel>>(projects);
        return new ApiResponse<IEnumerable<ProjectViewModel>>
        {
            Data = result
        };
    }

    [HttpGet]
    [Route("")]
    public ApiResponse<ProjectViewModel> GetProject([FromQuery] string projectId)
    {
        var project = projectService.GetProject(projectId);
        var result = mapper.Map<ProjectViewModel>(project);
        return new ApiResponse<ProjectViewModel>
        {
            Data = result
        };
    }

    [HttpPost]
    [Route("")]
    public ApiResponse<ProjectViewModel> AddProject([FromBody] AddProjectRequestModel req)
    {
        if (req.Name == null)
        {
            throw new WellKnownApiException("Name field is null", "INVALID_PARAMETER");
        }

        var project = projectService.AddProject(req.Name);
        var result = mapper.Map<ProjectViewModel>(project);
        return new ApiResponse<ProjectViewModel>
        {
            Data = result
        };
    }

    [HttpPut]
    [Route("")]
    public ApiResponse<ProjectViewModel> UpdateProject([FromBody] UpdateProjectRequestModel req)
    {
        if (req.ProjectId == null)
        {
            throw new WellKnownApiException("ProjectId field is null", "INVALID_PARAMETER");
        }
        if (req.Name == null)
        {
            throw new WellKnownApiException("Name field is null", "INVALID_PARAMETER");
        }

        var project = projectService.UpdateProject(req.ProjectId, req.Name);
        var result = mapper.Map<ProjectViewModel>(project);
        return new ApiResponse<ProjectViewModel>
        {
            Data = result
        };
    }

    [HttpDelete]
    [Route("")]
    public ApiResponse DeleteProject([FromQuery] string projectId)
    {
        projectService.DeleteProject(projectId);
        return new ApiResponse();
    }
}
